### PR TITLE
vim-patch:8.2.{3492,3493,3570,3573,3574,3575,3577,3601}: put overflow checking shenanigans

### LINKED ISSUE #12907: using ctrl-p after yanking more than 512 lines trigger undefined-behavior at ../src/nvim/ops.c:5441:17
- `nvim --version`:

```
NVIM v0.5.0-693-g20f7b9428
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /home/wangxinyu/.local/bin/clang -O2 -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -fno-sanitize-recover=all -fno-omit-frame-pointer -fno-optimize-sibling-calls -fsanitize=address -fsanitize=undefined -fsanitize-blacklist=/home/wangxinyu/neovim/src/.asan-blacklist -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -DEXITFREE -I/home/wangxinyu/neovim/build/config -I/home/wangxinyu/neovim/src -I/usr/include -I/home/wangxinyu/neovim/.deps/usr/include -I/home/wangxinyu/neovim/build/src/nvim/auto -I/home/wangxinyu/neovim/build/include
Compiled by wangxinyu@RESEARCH

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/home/wangxinyu/.local/share/nvim"

Run :checkhealth for more info
```

- `vim -u DEFAULTS` (version: ) behaves differently?  not tried
- Operating system/version:  archlinux 64bit
- Terminal name/version: konsole
- `$TERM`: screen   (using tmux)

### Steps to reproduce using `nvim -u NORC`

```
can not produce since crash must be triggered by Ctrl-P plugin
```

### Actual behaviour

1. create a file more than 512 lines ```seq 1 513 > large.txt```
2. edit file by nvim large.txt by ``` nvim large.txt```
3. enter visual mode and select from top left to bottom right  ``` ctrl-v => G => $ => y```
4. press ctrl-p , then nvim crash

asan log:
```
../src/nvim/ops.c:5441:17: runtime error: null pointer passed as argument 2, which is declared to never be null
/usr/include/string.h:43:28: note: nonnull attribute specified here
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior ../src/nvim/ops.c:5441:17 in
```

full debug backtrace:
```
#0  0x00007efc2eeacf25 in raise () from /usr/lib/libc.so.6
#1  0x00007efc2ee96897 in abort () from /usr/lib/libc.so.6
#2  0x0000000000526c6b in __sanitizer::Abort () at /home/wangxinyu/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_posix_libcdep.cpp:155
#3  0x0000000000525078 in __sanitizer::Die () at /home/wangxinyu/llvm-project/compiler-rt/lib/sanitizer_common/sanitizer_termination.cpp:58
#4  0x000000000053a635 in __ubsan::__ubsan_handle_nonnull_arg_abort (Data=<optimized out>) at /home/wangxinyu/llvm-project/compiler-rt/lib/ubsan/ubsan_handlers.cpp:719
#5  0x00000000009c2d2a in str_to_reg (y_ptr=<optimized out>, yank_type=yank_type@entry=kMTBlockWise, str=<optimized out>, len=len@entry=1943, blocklen=blocklen@entry=2, str_list=<optimized out>, str_list@entry=false) at ../src/nvim/ops.c:5441
#6  0x00000000009c1e06 in write_reg_contents_ex (name=<optimized out>, str=<optimized out>, len=1943, must_append=<optimized out>, yank_type=<optimized out>, block_len=<optimized out>) at ../src/nvim/ops.c:5356
#7  0x00000000006bd5cb in f_setreg (argvars=<optimized out>, rettv=<optimized out>, fptr=<optimized out>) at ../src/nvim/eval/funcs.c:8490
#8  0x00000000007140e2 in call_func (funcname=<optimized out>, len=<optimized out>, rettv=<optimized out>, argcount_in=<optimized out>, argvars_in=<optimized out>, argv_func=<optimized out>, firstline=<optimized out>, lastline=<optimized out>, 
    doesrange=<optimized out>, evaluate=<optimized out>, partial=<optimized out>, selfdict_in=<optimized out>) at ../src/nvim/eval/userfunc.c:1459
#9  0x00000000007132f6 in get_func_tv (name=<optimized out>, len=<optimized out>, rettv=<optimized out>, arg=0x7ffd0aef3180, firstline=<optimized out>, lastline=<optimized out>, doesrange=<optimized out>, evaluate=<optimized out>, partial=<optimized out>, 
    selfdict=<optimized out>) at ../src/nvim/eval/userfunc.c:422
#10 0x00000000007233aa in ex_call (eap=<optimized out>) at ../src/nvim/eval/userfunc.c:2856
#11 0x000000000077e7a8 in do_one_cmd (cmdlinep=<optimized out>, flags=<optimized out>, cstack=<optimized out>, fgetline=<optimized out>, cookie=<optimized out>) at ../src/nvim/ex_docmd.c:1949
#12 0x0000000000776434 in do_cmdline (cmdline=<optimized out>, fgetline=<optimized out>, cookie=<optimized out>, flags=<optimized out>) at ../src/nvim/ex_docmd.c:589
#13 0x0000000000715e5a in call_user_func (fp=<optimized out>, argcount=<optimized out>, argvars=<optimized out>, rettv=<optimized out>, firstline=<optimized out>, lastline=<optimized out>, selfdict=<optimized out>) at ../src/nvim/eval/userfunc.c:1038
#14 0x0000000000714127 in call_func (funcname=<optimized out>, len=<optimized out>, rettv=<optimized out>, argcount_in=<optimized out>, argvars_in=<optimized out>, argv_func=<optimized out>, firstline=<optimized out>, lastline=<optimized out>, 
    doesrange=<optimized out>, evaluate=<optimized out>, partial=<optimized out>, selfdict_in=<optimized out>) at ../src/nvim/eval/userfunc.c:1444
#15 0x00000000007132f6 in get_func_tv (name=<optimized out>, len=<optimized out>, rettv=<optimized out>, arg=0x7ffd0aef6960, firstline=<optimized out>, lastline=<optimized out>, doesrange=<optimized out>, evaluate=<optimized out>, partial=<optimized out>, 
    selfdict=<optimized out>) at ../src/nvim/eval/userfunc.c:422
#16 0x0000000000696053 in eval7 (arg=<optimized out>, rettv=<optimized out>, evaluate=<optimized out>, want_string=<optimized out>) at ../src/nvim/eval.c:4087
#17 0x0000000000693141 in eval6 (arg=<optimized out>, rettv=<optimized out>, evaluate=<optimized out>, want_string=<optimized out>) at ../src/nvim/eval.c:3797
#18 0x0000000000692457 in eval5 (arg=0x7ffd0aef6960, rettv=0x7ffd0aef5cc0, evaluate=1) at ../src/nvim/eval.c:3645
#19 0x000000000069126c in eval4 (arg=0x7ffd0aef6960, rettv=0x7ffd0aef5cc0, evaluate=1) at ../src/nvim/eval.c:3386
#20 0x0000000000690cdb in eval3 (arg=<optimized out>, rettv=<optimized out>, evaluate=<optimized out>) at ../src/nvim/eval.c:3304
#21 0x0000000000664e11 in eval2 (arg=0x7ffd0aef6960, rettv=0x7ffd0aef5cc0, evaluate=1) at ../src/nvim/eval.c:3235
#22 eval1 (arg=<optimized out>, rettv=<optimized out>, evaluate=<optimized out>) at ../src/nvim/eval.c:3161
#23 0x0000000000695122 in get_list_tv (arg=0x7ffd0aef6960, rettv=0x7ffd0aef6aa0, evaluate=1) at ../src/nvim/eval.c:4744
#24 eval7 (arg=<optimized out>, rettv=0x7ffd0aef6aa0, evaluate=<optimized out>, want_string=<optimized out>) at ../src/nvim/eval.c:4012
#25 0x0000000000693141 in eval6 (arg=<optimized out>, rettv=<optimized out>, evaluate=<optimized out>, want_string=<optimized out>) at ../src/nvim/eval.c:3797
#26 0x0000000000692457 in eval5 (arg=0x7ffd0aef6960, rettv=0x7ffd0aef6aa0, evaluate=1) at ../src/nvim/eval.c:3645
#27 0x000000000069126c in eval4 (arg=0x7ffd0aef6960, rettv=0x7ffd0aef6aa0, evaluate=1) at ../src/nvim/eval.c:3386
#28 0x0000000000690cdb in eval3 (arg=<optimized out>, rettv=<optimized out>, evaluate=<optimized out>) at ../src/nvim/eval.c:3304
#29 0x0000000000664e11 in eval2 (arg=0x7ffd0aef6960, rettv=0x7ffd0aef6aa0, evaluate=1) at ../src/nvim/eval.c:3235
#30 eval1 (arg=<optimized out>, rettv=<optimized out>, evaluate=<optimized out>) at ../src/nvim/eval.c:3161
#31 0x0000000000663ec1 in eval0 (arg=0x6060002f28ff "[bufnr('%'), s:lastvisual()]", rettv=<optimized out>, nextcmd=<optimized out>, evaluate=<optimized out>) at ../src/nvim/eval.c:3121
#32 0x0000000000667ecd in ex_let_const (eap=0x7ffd0aef6d60, is_const=<optimized out>) at ../src/nvim/eval.c:1420
#33 0x000000000066954b in ex_let (eap=0x2) at ../src/nvim/eval.c:1348
#34 0x000000000077e7a8 in do_one_cmd (cmdlinep=<optimized out>, flags=<optimized out>, cstack=<optimized out>, fgetline=<optimized out>, cookie=<optimized out>) at ../src/nvim/ex_docmd.c:1949
#35 0x0000000000776434 in do_cmdline (cmdline=<optimized out>, fgetline=<optimized out>, cookie=<optimized out>, flags=<optimized out>) at ../src/nvim/ex_docmd.c:589
#36 0x0000000000715e5a in call_user_func (fp=<optimized out>, argcount=<optimized out>, argvars=<optimized out>, rettv=<optimized out>, firstline=<optimized out>, lastline=<optimized out>, selfdict=<optimized out>) at ../src/nvim/eval/userfunc.c:1038
#37 0x0000000000714127 in call_func (funcname=<optimized out>, len=<optimized out>, rettv=<optimized out>, argcount_in=<optimized out>, argvars_in=<optimized out>, argv_func=<optimized out>, firstline=<optimized out>, lastline=<optimized out>, 
    doesrange=<optimized out>, evaluate=<optimized out>, partial=<optimized out>, selfdict_in=<optimized out>) at ../src/nvim/eval/userfunc.c:1444
#38 0x00000000007132f6 in get_func_tv (name=<optimized out>, len=<optimized out>, rettv=<optimized out>, arg=0x7ffd0aef88e0, firstline=<optimized out>, lastline=<optimized out>, doesrange=<optimized out>, evaluate=<optimized out>, partial=<optimized out>, 
    selfdict=<optimized out>) at ../src/nvim/eval/userfunc.c:422
#39 0x00000000007233aa in ex_call (eap=<optimized out>) at ../src/nvim/eval/userfunc.c:2856
#40 0x000000000077e7a8 in do_one_cmd (cmdlinep=<optimized out>, flags=<optimized out>, cstack=<optimized out>, fgetline=<optimized out>, cookie=<optimized out>) at ../src/nvim/ex_docmd.c:1949
#41 0x0000000000776434 in do_cmdline (cmdline=<optimized out>, fgetline=<optimized out>, cookie=<optimized out>, flags=<optimized out>) at ../src/nvim/ex_docmd.c:589
#42 0x0000000000715e5a in call_user_func (fp=<optimized out>, argcount=<optimized out>, argvars=<optimized out>, rettv=<optimized out>, firstline=<optimized out>, lastline=<optimized out>, selfdict=<optimized out>) at ../src/nvim/eval/userfunc.c:1038
#43 0x0000000000714127 in call_func (funcname=<optimized out>, len=<optimized out>, rettv=<optimized out>, argcount_in=<optimized out>, argvars_in=<optimized out>, argv_func=<optimized out>, firstline=<optimized out>, lastline=<optimized out>, 
    doesrange=<optimized out>, evaluate=<optimized out>, partial=<optimized out>, selfdict_in=<optimized out>) at ../src/nvim/eval/userfunc.c:1444
#44 0x00000000007132f6 in get_func_tv (name=<optimized out>, len=<optimized out>, rettv=<optimized out>, arg=0x7ffd0aefa780, firstline=<optimized out>, lastline=<optimized out>, doesrange=<optimized out>, evaluate=<optimized out>, partial=<optimized out>, 
    selfdict=<optimized out>) at ../src/nvim/eval/userfunc.c:422
#45 0x00000000007233aa in ex_call (eap=<optimized out>) at ../src/nvim/eval/userfunc.c:2856
#46 0x000000000077e7a8 in do_one_cmd (cmdlinep=<optimized out>, flags=<optimized out>, cstack=<optimized out>, fgetline=<optimized out>, cookie=<optimized out>) at ../src/nvim/ex_docmd.c:1949
#47 0x0000000000776434 in do_cmdline (cmdline=<optimized out>, fgetline=<optimized out>, cookie=<optimized out>, flags=<optimized out>) at ../src/nvim/ex_docmd.c:589
#48 0x0000000000715e5a in call_user_func (fp=<optimized out>, argcount=<optimized out>, argvars=<optimized out>, rettv=<optimized out>, firstline=<optimized out>, lastline=<optimized out>, selfdict=<optimized out>) at ../src/nvim/eval/userfunc.c:1038
#49 0x0000000000714127 in call_func (funcname=<optimized out>, len=<optimized out>, rettv=<optimized out>, argcount_in=<optimized out>, argvars_in=<optimized out>, argv_func=<optimized out>, firstline=<optimized out>, lastline=<optimized out>, 
    doesrange=<optimized out>, evaluate=<optimized out>, partial=<optimized out>, selfdict_in=<optimized out>) at ../src/nvim/eval/userfunc.c:1444
#50 0x00000000007132f6 in get_func_tv (name=<optimized out>, len=<optimized out>, rettv=<optimized out>, arg=0x7ffd0aefc620, firstline=<optimized out>, lastline=<optimized out>, doesrange=<optimized out>, evaluate=<optimized out>, partial=<optimized out>, 
    selfdict=<optimized out>) at ../src/nvim/eval/userfunc.c:422
#51 0x00000000007233aa in ex_call (eap=<optimized out>) at ../src/nvim/eval/userfunc.c:2856
#52 0x000000000077e7a8 in do_one_cmd (cmdlinep=<optimized out>, flags=<optimized out>, cstack=<optimized out>, fgetline=<optimized out>, cookie=<optimized out>) at ../src/nvim/ex_docmd.c:1949
#53 0x0000000000776434 in do_cmdline (cmdline=<optimized out>, fgetline=<optimized out>, cookie=<optimized out>, flags=<optimized out>) at ../src/nvim/ex_docmd.c:589
#54 0x0000000000780f1f in do_ucmd (eap=0x7ffd0aefd800) at ../src/nvim/ex_docmd.c:5987
#55 do_one_cmd (cmdlinep=<optimized out>, flags=<optimized out>, cstack=<optimized out>, fgetline=<optimized out>, cookie=<optimized out>) at ../src/nvim/ex_docmd.c:1943
#56 0x0000000000776434 in do_cmdline (cmdline=<optimized out>, fgetline=<optimized out>, cookie=<optimized out>, flags=<optimized out>) at ../src/nvim/ex_docmd.c:589
#57 0x0000000000980064 in nv_colon (cap=0x7ffd0aefe7c8) at ../src/nvim/normal.c:4625
#58 0x0000000000975636 in normal_execute (state=<optimized out>, key=<optimized out>) at ../src/nvim/normal.c:1138
#59 0x0000000000b828de in state_enter (s=s@entry=0x7ffd0aefe740) at ../src/nvim/state.c:69
#60 0x000000000095e8f9 in normal_enter (cmdwin=false, noexmode=<optimized out>) at ../src/nvim/normal.c:463
#61 0x00000000008a8781 in main (argc=<optimized out>, argv=<optimized out>) at ../src/nvim/main.c:557
```

this only happen if yank more than 512 lines, if yank less than 512 lines, nvim won't crash

### PR DESCRIPTION
**Pull request by seandewar**
Also fix #12907

### CODE REVIEW

**Comment by seandewar** [comment]
```
2022-02-16T23:39:03.5547553Z /home/runner/work/neovim/neovim/src/nvim/ops.c:5714:17: runtime error: null pointer passed as argument 2, which is declared to never be null
2022-02-16T23:39:03.5548057Z /usr/include/string.h:44:28: note: nonnull attribute specified here
2022-02-16T23:39:03.5548452Z     #0 0x133d3b7 in str_to_reg /home/runner/work/neovim/neovim/src/nvim/ops.c:5714:7
2022-02-16T23:39:03.5548865Z     #1 0x133b4fe in write_reg_contents_ex /home/runner/work/neovim/neovim/src/nvim/ops.c:5625:3
2022-02-16T23:39:03.5549264Z     #2 0x133ab5b in write_reg_contents /home/runner/work/neovim/neovim/src/nvim/ops.c:5508:3
2022-02-16T23:39:03.5550048Z     #3 0x11933c5 in redir_write /home/runner/work/neovim/neovim/src/nvim/message.c:3254:7
2022-02-16T23:39:03.5550472Z     #4 0x1194202 in msg_puts_attr_len /home/runner/work/neovim/neovim/src/nvim/message.c:1995:3
2022-02-16T23:39:03.5550876Z     #5 0x1193c32 in msg_puts_attr /home/runner/work/neovim/neovim/src/nvim/message.c:1982:3
2022-02-16T23:39:03.5551277Z     #6 0x118ed36 in msg_puts /home/runner/work/neovim/neovim/src/nvim/message.c:1945:3
2022-02-16T23:39:03.5551671Z     #7 0xbb8771 in list_func_head /home/runner/work/neovim/neovim/src/nvim/eval/userfunc.c:1639:3
2022-02-16T23:39:03.5552093Z     #8 0xba7797 in ex_function /home/runner/work/neovim/neovim/src/nvim/eval/userfunc.c:1988:15
2022-02-16T23:39:03.5552487Z     #9 0xcac62f in do_one_cmd /home/runner/work/neovim/neovim/src/nvim/ex_docmd.c:1975:5
2022-02-16T23:39:03.5552887Z     #10 0xc96683 in do_cmdline /home/runner/work/neovim/neovim/src/nvim/ex_docmd.c:594:20
2022-02-16T23:39:03.5553279Z     #11 0xc80ec2 in do_source /home/runner/work/neovim/neovim/src/nvim/ex_cmds2.c:2108:5
2022-02-16T23:39:03.5553667Z     #12 0xc7c475 in cmd_source /home/runner/work/neovim/neovim/src/nvim/ex_cmds2.c:1671:14
2022-02-16T23:39:03.5554068Z     #13 0xc7c553 in ex_source /home/runner/work/neovim/neovim/src/nvim/ex_cmds2.c:1652:3
2022-02-16T23:39:03.5554451Z     #14 0xcac62f in do_one_cmd /home/runner/work/neovim/neovim/src/nvim/ex_docmd.c:1975:5
2022-02-16T23:39:03.5554847Z     #15 0xc96683 in do_cmdline /home/runner/work/neovim/neovim/src/nvim/ex_docmd.c:594:20
2022-02-16T23:39:03.5555244Z     #16 0xc9c080 in do_cmdline_cmd /home/runner/work/neovim/neovim/src/nvim/ex_docmd.c:278:10
2022-02-16T23:39:03.5555631Z     #17 0xffb8b0 in exe_commands /home/runner/work/neovim/neovim/src/nvim/main.c:1708:5
2022-02-16T23:39:03.5556019Z     #18 0xfeae7a in main /home/runner/work/neovim/neovim/src/nvim/main.c:503:5
2022-02-16T23:39:03.5556829Z     #19 0x7fc6a32be0b2 in __libc_start_main /build/glibc-eX1tMB/glibc-2.31/csu/../csu/libc-start.c:308:16
2022-02-16T23:39:03.5557161Z     #0 0x45efdd in _start ??:?
2022-02-16T23:39:03.5557314Z 
2022-02-16T23:39:03.5557742Z SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /home/runner/work/neovim/neovim/src/nvim/ops.c:5714:17 in
2022-02-16T23:39:03.5558199Z Failed: E /oldtests|logs :: Runtime errors detected.
```

This failure is weird, it seems to not be triggered by this PR's code changes.
Reverting all the code changes from this PR but applying the test changes still triggers it.

Interestingly, commenting out any test seems to make it disappear. 🤔

**Comment by seandewar** [comment]
Looks like https://github.com/neovim/neovim/issues/12907...

**Comment by seandewar** [comment]
Looks like that works; I'll finish this off later.

File: src/nvim/testdir/test_put.vim [hunk 1]
```diff
@@ -139,10 +139,18 @@ func Test_p_with_count_leaves_mark_at_end()
...
+  bwipe!
+endfunc
 
-  if v:sizeofint != 8
-    throw 'Skipped: only works with 64 bit ints'
+func Test_very_large_count_64bit()
+  " v:sizeoflong is N/A; check for Windows instead
+  " if v:sizeoflong < 8
+  if has('win32')
+   throw 'Skipped: only works with 64 bit long ints'
```
Comment (zeertzjq) [inline src/nvim/testdir/test_put.vim:153]: indent

**Review by zeertzjq — COMMENTED** [review]
Is it really good to assume size of long by operating system?

**Comment by seandewar** [comment]
Maybe not. I'm not fully decided on whether or not to port the vimvars yet.

File: src/nvim/testdir/test_put.vim (hunk 1 above)
Comment (seandewar) [reply src/nvim/testdir/test_put.vim:153]: Thanks!

**Comment by seandewar** [comment]
Didn't consider the option of using the LuaJIT FFI (thanks bjorn). That seems to work.
I think this is ready now?

**Review by zeertzjq — APPROVED** [review]

**Pull request merged**